src/SqlPersistence/Config/SqlPersistenceConfig_Connection.cs

```diff
@@ -15,24 +16,70 @@ public static void ConnectionBuilder(this PersistenceExtensions<SqlPersistence>
+
+                return null;
+            });
+
+            MultiTenantConnectionBuilder(configuration, captureTenantId, buildConnectionFromTenantData);
+        }
+
+        // Possible future API, or if IReadOnlyDictionary<string, string> is replaced by a message headers abstraction
+        static void MultiTenantConnectionBuilder(this PersistenceExtensions<SqlPersistence> configuration, Func<IReadOnlyDictionary<string, string>, string> captureTenantId, Func<string, DbConnection> buildConnectionFromTenantData)
+        {
+            Guard.AgainstNull(nameof(configuration), configuration);
+            Guard.AgainstNull(nameof(captureTenantId), captureTenantId);
+            Guard.AgainstNull(nameof(buildConnectionFromTenantData), buildConnectionFromTenantData);
+
+            var connectionManager = new MultiTenantConnectionManager(captureTenantId, buildConnectionFromTenantData);
+
+            var settings = configuration.GetSettings();
+            settings.Set($"SqlPersistence.ConnectionManager.{typeof(StorageType.Outbox).Name}", connectionManager);
+            settings.Set($"SqlPersistence.ConnectionManager.{typeof(StorageType.Sagas).Name}", connectionManager);
+            settings.Set("SqlPersistence.MultiTenant", true);
         }
 
-        internal static Func<DbConnection> GetConnectionBuilder(this ReadOnlySettings settings, Type storageType)
+        internal static ConnectionManager GetConnectionBuilder(this ReadOnlySettings settings, Type storageType)
         {
-            if (settings.TryGet($"SqlPersistence.ConnectionBuilder.{storageType.Name}", out Func<DbConnection> value))
+            if (settings.TryGet($"SqlPersistence.ConnectionManager.{storageType.Name}", out ConnectionManager value))
             {
                 return value;
             }
-            if (settings.TryGet("SqlPersistence.ConnectionBuilder", out value))
+            if (settings.TryGet("SqlPersistence.ConnectionManager", out value))
             {
                 return value;
             }
-            throw new Exception($"Couldn't find connection string for {storageType}. The connection to the database must be specified using the ConnectionBuilder method.");
+
+            var exceptionMessage = $"Couldn't find connection string for {storageType.Name}. The connection to the database must be specified using the `{nameof(ConnectionBuilder)}` method.";
+
+            if (settings.EndpointIsMultiTenant())
+            {
+                exceptionMessage += $" When in multi-tenant mode with `{nameof(MultiTenantConnectionBuilder)}`, you must still use `{nameof(ConnectionBuilder)}` to provide a database connection for subscriptions/timeouts on message transports that don't support those features natively.";
+            }
+
+            throw new Exception(exceptionMessage);
         }
 
-        internal static Func<DbConnection> GetConnectionBuilder<T>(this ReadOnlySettings settings)
+        // TODO: RENAME to GetConnectionManager
```

Comment:
I think that given the fact that the connection manager only creates/opens the connection, the name `ConnectionBuilder` would be more appropriate.